backend/gateway/meta/meta.go

```diff
@@ -119,6 +119,77 @@ func ResourceNames(pb proto.Message) []*auditv1.Resource {
 	return nil
 }
 
+// ExtractProtoPatternsValues takes a proto and returns its pattern populated with values
+func ExtractProtoPatternsValues(pb proto.Message) string {
+	m := pb.ProtoReflect()
+	opts := m.Descriptor().Options().ProtoReflect()
+
+	populatedPattern := []string{}
+
+	if opts.Has(identifierTypeDescriptor) {
+		v := opts.Get(identifierTypeDescriptor)
+		id := v.Message().Interface().(*apiv1.Identifier)
+
+		for _, pattern := range id.Patterns {
+			rs := resolvePattern(pb, pattern)
+			populatedPattern = append(populatedPattern, rs.Id)
+		}
+	}
+	return populatedPattern[0]
+}
+
+// PatternValueMapping takes a string value and maps the patterns from a proto pattern
+// this is utilized by the resolver search api
+//
+// For example given the following proto pattern
+// option (clutch.api.v1.id).patterns = {
+//  pattern : "{cluster}/{namespace}/{name}"
+// };
+//
+// And the value of "mycluster/mynamespace/nameofresource"
+// we transform the pattern into a regex and map the values to the pattern names
+//
+// The output for this example is:
+// map[string]string{
+//  cluster: mycluster
+//  namespace: mynamespace
+//  name: nameofresource
+// }
+func PatternValueMapping(pb proto.Message, value string) (map[string]string, error) {
+	m := pb.ProtoReflect()
+	opts := m.Descriptor().Options().ProtoReflect()
+
+	// Field and Value result map
+	result := map[string]string{}
+
+	if opts.Has(identifierTypeDescriptor) {
+		v := opts.Get(identifierTypeDescriptor)
+		id := v.Message().Interface().(*apiv1.Identifier)
+
+		for _, pattern := range id.Patterns {
+			// The variable names on the pattern
+			variableNames := fieldNameRegexp.FindAllStringSubmatch(pattern.Pattern, -1)
```

Comment:
i would pull this out into a func that returns `[]string` and use everywhere we're doing this

Reply:
Looking at this more both use cases for this have slightly different needs, i would need to return the raw output to satisfy both. Leaving this func with just a single line.

```
func extractProtoPatternFieldNames(pattern *apiv1.Pattern) [][]string {
	return  fieldNameRegexp.FindAllStringSubmatch(pattern.Pattern, -1)
}
```

`resolverPattern` is the other usage 
https://github.com/lyft/clutch/blob/f5e5ce7ea88be003e99dcc1fffffcabe5401ad6a/backend/gateway/meta/meta.go#L246